# coding: utf-8

"""
    ODM API

    This swagger page describes the variantUser API endpoints for ODM. These are typically used to find and retrieve variant data and metadata.  Before carrying out any API calls you will need an API token. API tokens can be obtained under your profile within the Genestack software.  To try out calls in this swagger page:  1.  Click the 'Authorize' button below to enter your API token 2.  Scroll to the 'Parameters' section for the method you wish to try out and click the 'Try it out' button 3.  Enter parameter values that you wish to try 4.  Scroll to the bottom of the Parameters section and click the 'Execute' bar that appears    The server response will be in the section that follows.

    The version of the OpenAPI document: default-released
    Generated by OpenAPI Generator (https://openapi-generator.tech)

    Do not edit the class manually.
"""  # noqa: E501


from __future__ import annotations
import pprint
import re  # noqa: F401
import json

from pydantic import BaseModel, ConfigDict, Field, StrictBool, StrictInt, StrictStr
from typing import Any, ClassVar, Dict, List, Optional
from odm_api.models.find_objects_response_objects_page_content_inner_summary_inner import FindObjectsResponseObjectsPageContentInnerSummaryInner
from typing import Optional, Set
from typing_extensions import Self

class StudySearchInfo(BaseModel):
    """
    StudySearchInfo
    """ # noqa: E501
    accession: Optional[StrictStr] = None
    name: Optional[StrictStr] = None
    owner: Optional[StrictStr] = None
    file_creation: Optional[StrictStr] = Field(default=None, alias="fileCreation")
    size: Optional[StrictInt] = None
    has_facs: Optional[StrictBool] = Field(default=None, alias="hasFacs")
    has_genomic: Optional[StrictBool] = Field(default=None, alias="hasGenomic")
    has_transcriptomics: Optional[StrictBool] = Field(default=None, alias="hasTranscriptomics")
    summary: Optional[List[FindObjectsResponseObjectsPageContentInnerSummaryInner]] = None
    __properties: ClassVar[List[str]] = ["accession", "name", "owner", "fileCreation", "size", "hasFacs", "hasGenomic", "hasTranscriptomics", "summary"]

    model_config = ConfigDict(
        populate_by_name=True,
        validate_assignment=True,
        protected_namespaces=(),
    )


    def to_str(self) -> str:
        """Returns the string representation of the model using alias"""
        return pprint.pformat(self.model_dump(by_alias=True))

    def to_json(self) -> str:
        """Returns the JSON representation of the model using alias"""
        # TODO: pydantic v2: use .model_dump_json(by_alias=True, exclude_unset=True) instead
        return json.dumps(self.to_dict())

    @classmethod
    def from_json(cls, json_str: str) -> Optional[Self]:
        """Create an instance of StudySearchInfo from a JSON string"""
        return cls.from_dict(json.loads(json_str))

    def to_dict(self) -> Dict[str, Any]:
        """Return the dictionary representation of the model using alias.

        This has the following differences from calling pydantic's
        `self.model_dump(by_alias=True)`:

        * `None` is only added to the output dict for nullable fields that
          were set at model initialization. Other fields with value `None`
          are ignored.
        """
        excluded_fields: Set[str] = set([
        ])

        _dict = self.model_dump(
            by_alias=True,
            exclude=excluded_fields,
            exclude_none=True,
        )
        # override the default output from pydantic by calling `to_dict()` of each item in summary (list)
        _items = []
        if self.summary:
            for _item_summary in self.summary:
                if _item_summary:
                    _items.append(_item_summary.to_dict())
            _dict['summary'] = _items
        return _dict

    @classmethod
    def from_dict(cls, obj: Optional[Dict[str, Any]]) -> Optional[Self]:
        """Create an instance of StudySearchInfo from a dict"""
        if obj is None:
            return None

        if not isinstance(obj, dict):
            return cls.model_validate(obj)

        _obj = cls.model_validate({
            "accession": obj.get("accession"),
            "name": obj.get("name"),
            "owner": obj.get("owner"),
            "fileCreation": obj.get("fileCreation"),
            "size": obj.get("size"),
            "hasFacs": obj.get("hasFacs"),
            "hasGenomic": obj.get("hasGenomic"),
            "hasTranscriptomics": obj.get("hasTranscriptomics"),
            "summary": [FindObjectsResponseObjectsPageContentInnerSummaryInner.from_dict(_item) for _item in obj["summary"]] if obj.get("summary") is not None else None
        })
        return _obj


